Refactoring in goto_symex

romainbrenguier opened the pull request:
This are small refactoring commits in preparation for https://github.com/diffblue/cbmc/pull/3986

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Use namespace follow on type in rename(typet)

This avoids a disjunction of cases for the different type of tags and
makes the rename function simpler.

romainbrenguier committed:
Replace insert by emplace and UNREACHABLE by CHECK_RETURN

Using emplace avoids a call to make_pair, and CHECK_RETURN is more
adapted to the situation were we check the return value of insert/emplace.

romainbrenguier committed:
Make get_failed_symbol return optional symbol_exprt

This is more precise on the type of object returned.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f47879d8d72f049d4f92841f7619e4d02c3a8e53).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98997921
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c3bede40d1646353ca1a6df2c81f084bf7c7def3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99004545

martin-cs commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -690,18 +690,15 @@ void goto_symex_statet::rename(
   }
 
-  if(type.id()==ID_array)
+  if(const auto &array_type = type_try_dynamic_cast<array_typet>(type))
```
I guess this fits more with the current way of doing things, so, yes.

kroening replied:
I am no fan -- this doesn't improve execution speeds, but is less readable.

owen-mc-diffblue replied:
It provides extra validation beyond just checking the id, which depends on the type you are casting to

thomasspriggs replied:
I am in favour of using `type_try_dynamic_cast`, because it means that you only need to know the name of the type you want to cast to, you don't need to know that the `.id()` is supposed to be `ID_array` for an `array_typet`, in order to do the cast. It also reduces the likely-hood of mismatches between the check and the cast. There is little chance of checking for `ID_array` and casting to `bool_typet`, when both operations are done by the `try_cast`. I would however be in favour of renaming `type_try_dynamic_cast` to something less verbose.

martin-cs commented on src/goto-symex/symex_start_thread.cpp:
```diff
@@ -68,10 +68,9 @@ void goto_symext::symex_start_thread(statet &state)
-      UNREACHABLE;
+    auto emplace_result = state.level1.current_names.emplace(
+      lhs.get_l1_object_identifier(), std::make_pair(lhs, 0));
+    CHECK_RETURN(emplace_result.second);
```
Definitely.

martin-cs reviewed (commented):
The pointer-analysis changes seem fine.  I am unsure about the value of moving to the functional "lots of small static function" from the OO "protected utility functions".  This seems an issue of style and one that risks flip-flopping depending on who is working on the code.  I'm not sure here is the place to have that discussion.

peterschrammel reviewed (approved):
I'm in favour of small static functions for a class that is not going to be inherited from. It makes the class interface easier to read (even if it is a drop in the ocean in goto-symex).

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -261,13 +269,17 @@ static void set_l1_indices(
+static void set_l2_indices(
+  symex_level0t level0,
+  symex_level1t level1,
+  symex_level2t level2,
```
Can we please pass all these as const references? That will also show that their `operator()` should have been marked `const` (or maybe even `static`?), which will necessitate another commit to be added.

romainbrenguier replied:
Done and I've marked the operator const for symex_level1t, and replaced symex_level0t by a function so it doesn't need to be passed as an argument here.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -284,6 +284,25 @@ static void set_l2_indices(
+/// When the type of the operands of a `with_exprt` or `if_exprt` have been
+/// renamed they could end up being different from that of the expression. We
+/// fix that by propagating the type of its operands to \p expr.
+static void fix_type(exprt &expr)
```
I do note that I consider this a workaround for a genuine bug, and I'd be curious to see what happens if instead of replacing we actually checked that the types already are consistent?

kroening replied:
This should really not be necessary. We need to get rid of that renaming.

romainbrenguier replied:
Indeed it doesn't seem necessary. I've tested adding a invariant that checks that this does not do anything and all tests in regression/cbmc passed so there doesn't seem to be a bug we need to workaround. I will remove that part.

tautschnig replied:
As a matter of procedure, would you mind just dropping the commit from this PR, and create a separate PR that removes that code and replaces it by the invariant you've played with? Thank you!!

romainbrenguier replied:
made the PR there: https://github.com/diffblue/cbmc/pull/3993 and dropped the commit from here ~rebased this one on it~

tautschnig commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -32,14 +32,11 @@ void goto_symext::symex_dead(statet &state)
-    else
-      rhs=exprt(ID_invalid);
+    exprt rhs = [&]() -> exprt {
+      if(auto failed = get_failed_symbol(to_symbol_expr(code.symbol()), ns))
```
The `to_symbol_expr` is actually unnecessary, `code.symbol()` already returns a `symbol_exprt`.

romainbrenguier replied:
done

kroening reviewed (changes requested):
This needs another go -- I am in particularly concerned about the "renaming fix".

kroening commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -618,6 +644,20 @@ void goto_symex_statet::rename_address(
+    return ns.lookup(to_struct_tag_type(type)).type;
+  return {};
+}
+
```
This should not be necessary.

smowton replied:
If you mean "renaming is never needed for tagged types," I think that's untrue when we have structs with a flexible array member to rename.

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -27,6 +27,14 @@ Author: Daniel Kroening, kroening@kroening.com
 static void get_l1_name(exprt &expr);
 
+static void set_l2_indices(
+  symex_level0t level0,
```
Commented on the other PR, but can you explain the by-value passing here?

romainbrenguier replied:
the reference was omitted by mistake, but this reveals that this level0 is actually constant, I going to replace the symex_level0t by a function

romainbrenguier committed:
Mark symex_level0 and level1 operator() const

Make obvious the operator doesn't change the state of the object.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f5d3a86ddb8c4e4c02760be7c9cd2586af072fc8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99114133
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented:
@kroening Is this looking better?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 02c7e5b247e3f04ab086b8c796d9205c9afeacad).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99164643

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ce7a89854aa5aaad07f2afe57d1f27d0aa195baf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99238452

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b55e0aa33ab0fcbf9b7817102c8cb2447f292706).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99901122

kroening reviewed (changes requested):
* I don't see value in the static function refactoring; these do remind me too much of 90ies PASCAL.

* I am not a fan of the type_try_dynamic_cast idiom. At the very least it should be renamed, as 'dynamic_cast' has very specific meaning in C++.

* Two branches get replaced by a lambda; for most this reduces readability.

* The class hierarchy for the renaming levels is very intentional. I understand that the one that's replaced is trivial, but there's a point to the pattern.

* I am fine with the use of emplace instead of insert and the change of the signature of get_failed_symbol.

NathanJPhillips commented:
`dynamic_cast` in C++ is a checked downcast that relies on RTTI. `type_try_dynamic_cast` and friends in CBMC are checked downcasts of `irept` that rely on the ID field (or other signatures). The original design closely mirrored the C++ versions (one override for pointers that returned null on failure and the other for references that threw `std::bad_cast`). The primary advantage over the old version is locality of code - knowledge of the hard-coded ID's that correspond to the particular types is local to the definition of that type, users only need to know the name of the type they want to cast to. This increases code clarity at the call site.

LAJW commented:
> I don't see value in the static function refactoring; these do remind me too much of 90ies PASCAL.

Making member functions free removes `this`. Which removes all unnecessary implicit state. That way you know _exactly_ what the function is using without having to read through it.

If anything, to me the code before looks like 90ies Pascal 😁 (just replace `this` with global)

> Two branches get replaced by a lambda; for most this reduces readability.

These are "immediately invoked lambda expressions" aka closures or named scopes from the likes of OCaml/F#/Haskell:
```haskell
let value =
  let x = 5
  let y = 3
  x + y
-- value == 8, x and y are not accessible here
```

romainbrenguier commented:
@kroening 
> I don't see value in the static function refactoring; these do remind me too much of 90ies PASCAL.

I agree with @LAJW comment on that, plus it is generally recommended to make functions non-member whenever possible
(see http://www.drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197)
and it makes the other changes I want easier.

> I am not a fan of the type_try_dynamic_cast idiom. At the very least it should be renamed, as 'dynamic_cast' has very specific meaning in C++.

See @NathanJPhillips comment on that, and renaming it would be outside the scope of this PR.

> Two branches get replaced by a lambda; for most this reduces readability.

I removed these two lambda's.

> The class hierarchy for the renaming levels is very intentional. I understand that the one that's replaced is trivial, but there's a point to the pattern.

I removed this commit, though I don't see the value of this, the inheritance is not used, and it initializes a map that is never used.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: beb87d3a0ad533b498a84b75839b2fab3da0f004).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100460459

peterschrammel commented:
> I don't see value in the static function refactoring; these do remind me too much of 90ies PASCAL.

That could be considered a matter of taste. But, still, there are clear advantages of using static functions: 
* it declutters class declarations, makes them shorter and easier to read, and reduces the size of header files, speeding up compilation.
* it makes side effects explicit, making the code easier to understand, change and refactor (and trivial to refactor for reuse).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 569ecb651f34a87200b7a01c82eb3ae6fc012e37).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100766529

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -231,32 +239,41 @@ void goto_symex_statet::assignment(
 
-void goto_symex_statet::set_l0_indices(
+static void set_l0_indices(
+  symex_level0t &level0,
```
This can surely be const

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -725,22 +754,9 @@ void goto_symex_statet::rename(
-    rename(type, l1_identifier, ns, level);
-  }
-  else if(type.id() == ID_union_tag)
+  else if(const auto &followed_type = type_of_symbol_or_tag(type, ns))
```
This isn't a reference

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -664,6 +682,20 @@ static bool requires_renaming(const typet &type, const namespacet &ns)
+/// Return the type corresponding to a symbol or tag. Empty optional if \p type
+/// is neither a symbol or a tag.
+static optionalt<typet>
+type_of_symbol_or_tag(const typet &type, const namespacet &ns)
```
I think you could replace all this with `type = ns.follow(type);`, at the top of `rename_type`, as that's a no-op for non-tags.

smowton commented:
If I were you I'd drop the static-ifcation and have the `try_dynamic_cast` vs. `id() == ` fight elsewhere, that way this PR will be style-war free and easier to merge :)

romainbrenguier commented:
@smowton 
> If I were you I'd drop the static-ifcation and have the `try_dynamic_cast` vs. `id() == ` fight elsewhere, that way this PR will be style-war free and easier to merge :)

I've dropped the dynamic_cast change, and applied your suggestion for the `rename typet`, but making the functions static is the main purpose of this PR.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -688,6 +688,8 @@ void goto_symex_statet::rename(
   const namespacet &ns,
   levelt level)
 {
+  type = ns.follow(type);
```
I'm not even sure that compiles, but even when you make it compile: I think it should be done after `requires_renaming` to make sure we don't unnecessarily break sharing.

tautschnig commented on src/goto-symex/symex_start_thread.cpp:
```diff
@@ -68,10 +68,9 @@ void goto_symext::symex_start_thread(statet &state)
-                       std::make_pair(lhs, 0))).second)
-      UNREACHABLE;
+    auto emplace_result = state.level1.current_names.emplace(
+      lhs.get_l1_object_identifier(), std::make_pair(lhs, 0));
```
Until we have C++17 use:
````c++
auto emplace_result = state.level1.current_names.emplace(
  std::piecewise_construct,
  std::forward_as_tuple(lhs.get_l1_object_identifier()),
  std::forward_as_tuple(lhs, 0));
````
to avoid even constructing the second pair if the key already exists.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5e95d6937c1f1476fad6ff68d48eaf238fcd1921).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100977333
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5e95d6937c1f1476fad6ff68d48eaf238fcd1921).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100977333

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6bd85cfb9ebb10d27cedc6a0fc34fa05527397d1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101311049

kroening commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -675,6 +675,8 @@ void goto_symex_statet::rename(
     return; // no action
 
+  type = ns.follow(type);
+
```
Is that not expanding some tag types?

tautschnig replied:
Intentionally so, with the idea to replace the expansion below (lines 728ff have been removed). But I might be missing something.

kroening replied:
Ok -- I'd move below the following `if` block, and then add a comment such as
`// expand struct and union tag types`

smowton commented:
@romainbrenguier I talked to @kroening on Slack, he said this is ok once the `ns.follow` is moved below the `if(...)` immediately below it that may return the previously renamed type and therefore render the `follow` pointless.

romainbrenguier commented:
@smowton @kroening done

kroening commented:
@romainbrenguier Many thanks for the extra effort!
